pkg/generator/latch.go

```diff
@@ -0,0 +1,47 @@
+package generator
+
+import "sync"
+
+// ProtocolLatch increases the internal counter every time protocol execution
+// starts and decreases the counter every time protocol execution completes.
+// The latch implements Protocol interface and can be registered in the
+// Scheduler.
+//
+// The protocol code using the latch must guarantee that:
+// 1. `Lock()` is always called before `Unlock()`
+// 2. `Unlock()` is eventually called for every `Lock()`.
+//
+// Note that the Unlock() function may panic if the conditions are not met.
+type ProtocolLatch struct {
```

Comment:
Let's add docstrings for all public functions of this type.

Reply:
https://github.com/keep-network/keep-core/pull/3183/commits/c2a0003b35115557bee64dea9b8dee0d941fc030